src/main/java/org/testng/internal/thread/graph/GraphThreadPoolExecutor.java

```diff
@@ -61,20 +64,35 @@ public void run() {
    */
   private void runNodes(List<T> freeNodes) {
     List<IWorker<T>> runnables = m_factory.createWorkers(freeNodes);
+    CountDownLatch latch = null;
+    if (useLatches && runnables.size() > 1) {
+      latch = new CountDownLatch(runnables.size());
+    }
+    boolean latchUsed = false;
     for (IWorker<T> r : runnables) {
       m_activeRunnables.add(r);
+      if (latch != null && (r instanceof TestMethodWorker)) {
+        //Being conservative here and setting this only for test method workers
+        r.setLatch(latch);
+        latchUsed = true;
+      }
       ppp("Added to active runnable");
       setStatus(r, Status.RUNNING);
       ppp("Executing: " + r);
       try {
         execute(r);
-//        if (m_threadCount > 1) execute(r);
-//        else r.run();
       }
       catch(Exception ex) {
         ex.printStackTrace();
       }
     }
+    if (latch != null && latchUsed) {
+      try {
+        latch.await(10, TimeUnit.SECONDS);
```

Comment:
@juherr 

Please consider the test classes TestNG1, TestNG2, TestNG3 that are part of the the package [test.github1336](https://github.com/krmahadevan/testng/tree/8ad97dd15fe370cc272180455776cc19e73f8fbd/src/test/java/test/github1336).

For the first iteration : 
`org.testng.TestRunner#privateRun` > `org.testng.internal.thread.graph.GraphThreadPoolExecutor#run` > `org.testng.internal.thread.graph.GraphThreadPoolExecutor#runNodes`

The free-nodes would be as below : 

1. TestNG1.test1TestNG1()[pri:1, instance:test.github1336.TestNG1@23ab930d]
2. TestNG2.test1TestNG2()[pri:1, instance:test.github1336.TestNG2@4534b60d]
3. TestNG3.test1TestNG3()[pri:1, instance:test.github1336.TestNG3@3fa77460]

For the second iteration : 
`org.testng.internal.thread.graph.GraphThreadPoolExecutor#afterExecute` > `org.testng.internal.thread.graph.GraphThreadPoolExecutor#runNodes`

The free-nodes would be as below : 

Working with the freeNodes :
1. TestNG1.test2TestNG1()[pri:2, instance:test.github1336.TestNG1@23ab930d]
2. TestNG2.test2TestNG2()[pri:2, instance:test.github1336.TestNG2@4534b60d]
3. TestNG3.test2TestNG3()[pri:2, instance:test.github1336.TestNG3@3fa77460]
4. TestNG1.test3TestNG1()[pri:3, instance:test.github1336.TestNG1@23ab930d]
5. TestNG2.test3TestNG2()[pri:3, instance:test.github1336.TestNG2@4534b60d]
6. TestNG3.test3TestNG3()[pri:3, instance:test.github1336.TestNG3@3fa77460]


Without the latch, TestNG would immediately pick up the second set of free nodes without waiting for the first set of threads to run to completion.

Using an shutdown, should be done only after the all the threads have been submitted and not before that (its being done in `org.testng.internal.thread.graph.GraphThreadPoolExecutor#afterExecute` after the graph node has all nodes as completed ) and without invoking shutdown, there's no use of calling `awaitTermination`